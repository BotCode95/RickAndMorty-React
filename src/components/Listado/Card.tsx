import { useCard } from "../../hooks/useCard";
import { Result } from "../../interfaces/data";
import { Button } from "../Buttons/Button";

interface Props {
    character : Result
}

export const Card = ({character} : Props ) => {

  const {image,name, species, status, origin,location} : Result = character;
  const {favoriteStar,goBack,isFavoriteBtn} = useCard({character})
  return (
    <section className="row container container-card">
        <div className="col-1">
            <Button className="btn btn-back" onClick={goBack} >
                <i className="bi bi-arrow-left"></i>
            </Button>
        </div>
        <div className="col-4 img-container">
            <img src={image} className="img-character" alt={name}/>
        </div>
        <article className="col-7 article-card">
            <div className="row">
                <h3 className="col-10">{name}</h3>
                <div className="col-2 favorite">
                    <Button className="btn btn-info" onClick={isFavoriteBtn}>
                        {favoriteStar ? (
                            <i className="bi bi-star-fill favorite-icon"></i> 
                        ): (
                            <i className="bi bi-star"></i>
                        )}
                    </Button>
                </div>
            </div>
            <i className={`bi bi-circle-fill icon-status ${(status === 'Alive')  ? 'alive-icon' : 'dead-icon'}`}></i>
             {status} - {species}
            <div> 
                <span>First seen in:</span>
                <p>{origin.name}</p>
            </div>
            <div>
                <span>Last known location:</span>
                <p>{location.name}</p>
            </div>
        </article>
    </section>
    )
};
