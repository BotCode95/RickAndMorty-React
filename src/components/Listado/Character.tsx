import { Link } from "react-router-dom";
interface Props {
    id : number,
    name: string
}

export const Character = ({id,name} : Props ) => {
  return (
    <Link to={`/details/${id}`} className="item-list">
        {name}
    </Link>
  );
};
