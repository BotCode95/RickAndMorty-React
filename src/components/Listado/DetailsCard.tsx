import { useContext, useEffect } from "react";
import { Params, useParams } from "react-router-dom";
import { CharacterContext } from "../../context/Characters/CharacterContext";
import { Title } from "../Layout/Title";
import { Spinner } from "../Spinner/Spinner";
import { Card } from "./Card";


export const DetailsCard = () => {

    const {id} : Readonly<Params<string>>= useParams();
    const {cargando, character, getCharacterById} = useContext(CharacterContext);

    useEffect(() => {
        getCharacterById(parseInt(id!))
    // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [])
  return (
    <div>
        {cargando ? <Spinner/> : null}
        {
            character !== null && !cargando ? 
            <>
              <Title title={`Character : ${character?.name}`}/>
              <Card character={character!}/>
            </>
            : null
        }
    </div>
  );
};
