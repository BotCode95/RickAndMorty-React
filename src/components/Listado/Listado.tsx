import {useEffect, useContext, useRef} from "react";
import { CharacterContext } from "../../context/Characters/CharacterContext";
import { Card } from "./Card";

export const Listado = () => {
    const characterContext = useContext(CharacterContext);
    const {results, getCharacters} = characterContext;
    
    useEffect(() => {
        if(results?.length !== 0){
            // getCharactersByPage()
        }
        getCharacters()
        
    }, [])
  return (
    <>
    <div>
        <h2>Listado</h2>
    </div>
        <table>
            <thead>
                <tr>
                    <th>Name</th>
                    <th>Image</th>
                    <th>Status</th>
                    <th>Origin</th>
                    <th>Species</th>
                    <th>Location</th>
                </tr>
            </thead>
            {results?.map((item) : React.ReactElement => (
                
                <tbody key={item.id}>
                    <Card item={item}/>
                </tbody>
            ))}
        </table>
        
    </>
  )
};
