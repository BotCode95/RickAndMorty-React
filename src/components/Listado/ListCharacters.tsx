import { usePageList } from "../../hooks/usePageList";
import { Button } from "../Buttons/Button";
import { Spinner } from "../Spinner/Spinner";
import { Character } from "./Character";

export const ListCharacters = () => {
    const {cargando,finalPage,results,pageRef, pageNext, pagePrev} = usePageList();
  return (
    <>    
    {cargando ? <Spinner/> : null }    
    {results?.map((item) : React.ReactElement => (
        <div className="container container-listado" key={item.id}>
            <Character id={item.id} name={item.name}/>
        </div>
    ))}
        <div className="row justify-content-center">
            {pageRef.current !== 1 && (
                <Button 
                  className="btn btn-danger col-2 ms-2 mb-5 btn-pages" 
                  onClick={pagePrev} 
                >
                  Anterior
                </Button>
            )}
            {!finalPage && (
                <Button 
                    className="btn btn-info col-2 ms-2 mb-5 btn-pages" 
                    onClick={pageNext} 
                >
                    Siguiente
                </Button>
            )}
        </div>
    </>
  )
};
