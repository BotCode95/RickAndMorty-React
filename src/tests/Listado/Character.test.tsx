import React from 'react'
import {shallow} from 'enzyme'
import { Character } from '../../components/Listado/Character';

describe(('Test in <Character/>'), ()=> {
    const item = {
        "id": 1,
        "name": "Rick Sanchez",
    };
    const wrapper = shallow( <Character id={item.id} name={item.name}/> );
    
    test('should render component', () => {  

        expect(wrapper).toMatchSnapshot();
    });
})