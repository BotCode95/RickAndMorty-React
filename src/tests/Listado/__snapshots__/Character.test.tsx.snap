// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in <Character/> should render component 1`] = `
<Link
  className="item-list"
  to="/details/1"
>
  Rick Sanchez
</Link>
`;
